### DISCUSSION TITLE
[webpack 5] the third library of type=module will build failed

### DISCUSSION THREAD

**Discussion by kerm1it** [webpack/webpack#15211] [Q&A]
# Bug report

**What is the current behavior?**

when build the third library that `type` field is `module` of package.json，it won't work correctly if the code contains `require()`. (although this way is wrong)

But it will work correctly if I build it with webpack 4.

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

it works correctly in webpack 5.

**Other relevant information:**
webpack version:  5.66.0
Node.js version:  14.17.5
Operating System: mac
Additional tools: no

**Comment by alexander-akait** [comment]
Sorry, can you clarify how you use `require`?

**Reply by kerm1it** [reply]
``` js
let a;

if (process.env.PLATFORM === 'IOS') {
   a = require('./IOS/a.js');
} else {
   a = require('./Android/a.js');
}
```

**Reply by alexander-akait** [reply]
How you run this file?

**Reply by kerm1it** [reply]
build with webpack 5 and run it, but throw a error with `Can't find variable require`. webpack 4 is normal

**Reply by alexander-akait** [reply]
Do you have `target: node`?

**Reply by kerm1it** [reply]
No，I create a empty project with https://createapp.dev/webpack/no-library

**Reply by kerm1it** [reply]
This is `webpack.config.js`:
``` js
const webpack = require('webpack');
const path = require('path');

const config = {
  entry: './src/index.ts',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'bundle.js'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        use: 'babel-loader',
        exclude: /node_modules/
      },
      {
        test: /\.ts(x)?$/,
        loader: 'ts-loader',
        exclude: /node_modules/
      },
      {
        test: /\.svg$/,
        use: 'file-loader'
      },
      {
        test: /\.png$/,
        use: [
          {
            loader: 'url-loader',
            options: {
              mimetype: 'image/png'
            }
          }
        ]
      }
    ]
  },
  resolve: {
    extensions: [
      '.tsx',
      '.ts',
      '.js'
    ]
  }
};

module.exports = config;
```

**Reply by alexander-akait** [reply]
Because you have:
```
const webpack = require('webpack');
const path = require('path');
```

but say your in module format, change it on `import` and `module.exports` on `export default`

**Reply by kerm1it** [reply]
This isn't file of wrong code, it is only `webpack.config.js` to use build.

**Reply by kerm1it** [reply]
Let me describe the background again:

Previously, our project was packaged with webpack 4, which used a third-party library whose `type` field in package.json was `module`, but in order to achieve sub-platform packaging, code similar to the following was used:

``` js
let a;

if (process.env.PLATFORM === 'IOS') {
    a = require('./IOS/a.js');
} else {
    a = require('./Android/a.js');
}
````
It can be seen that the `require` method is used here to load the file, which can be packaged and run normally when packaged

Now that our project is migrated to webpack 5, after the packaging is completed, there is an error of `can't find require variable` when running

**Reply by alexander-akait** [reply]
Sorry, hard to say what is wrong, I need to look at configuration and code, maybe you can provide small example, no need to put all code, just provide example of the problem

**Reply by kerm1it** [reply]
[webpack5-require-reproduce.zip](https://github.com/webpack/webpack/files/7910215/webpack5-require-reproduce.zip)

This is a simple reproduce project.

View ` README.md` to reproduce after downloading.

**Reply by alexander-akait** [reply]
I can't run example due `TypeError: loaderContext.getOptions is not a function`, but yes webpack drop unnecessary `else` branch

**Reply by alexander-akait** [reply]
oh, I see, different in:
```
a = __webpack_require__(/*! ./b */ "../wrong-module/b.js");
```

and

```
a = require('./b');
```

**Reply by alexander-akait** [reply]
It is breaking change for webpack v5, because you use `"type": "module"`, webpack thinking code in ESM, but `require` doesn't exists in ESM code, so webpack is keeping code as is

**Reply by kerm1it** [reply]
No workaround or other solutions? because this library is third-party，I can't modify it.

**Reply by alexander-akait** [reply]
There is always workaround :smile: 

>  because this library is third-party，I can't modify it.

So `wrong-module` is third-party and you can't nothing to modify?

**Reply by kerm1it** [reply]
Yes

**Reply by alexander-akait** [reply]
Try this:
```
 {
        "type": "javascript/dynamic",
        test: /wrong-module/,
        loader: 'babel-loader',
        options: {
          "presets": [
            [
              "@babel/preset-env",
              {
                "modules": "commonjs"
              }
            ]
          ],
          sourceType: "unambiguous"
        }
      },
```

We say code is `unambiguous` (i.e. mix of `require`/`import`/`module.exports`/`export`) and convert it to commonjs
